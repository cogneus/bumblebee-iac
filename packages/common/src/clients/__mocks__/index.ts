export * from '../db/__mocks__'
export * from '../secrets/__mocks__'
export * from '../ssm/__mocks__'
