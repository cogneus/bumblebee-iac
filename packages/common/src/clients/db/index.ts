export * from './types'
export * from './db-client.class'
export * from './make-update-expression.helper'
export * from './get-record-limit.helper'
