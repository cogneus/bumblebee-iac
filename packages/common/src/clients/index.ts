export * from './db'
export * from './secrets'
export * from './ssm'
export * from './ensure-client'
export * from './types'
export * from './__mocks__'
export * from './destroy-clients'
