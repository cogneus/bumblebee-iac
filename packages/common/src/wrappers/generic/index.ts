export * from './generic.wrapper'
export * from './types'
