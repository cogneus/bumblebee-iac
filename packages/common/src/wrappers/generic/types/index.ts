export * from './generic-dependencies.interface'
export * from './generic-handler.type'
export * from './generic-wrapper.type'
export * from './generic-wrapper-params.interface'
