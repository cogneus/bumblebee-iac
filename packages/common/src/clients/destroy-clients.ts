import { Clients } from './types'

export const destroyClients = (clients: Clients) => {
  if (clients) {
    Object.values(clients).forEach((client) => client?.destroy())
  }
}
