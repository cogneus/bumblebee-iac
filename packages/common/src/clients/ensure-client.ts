import { InvalidService } from '../exceptions'
import { ServiceContext } from '../types'
import {
  ClientNames,
  Destroyable,
  SMClient,
  SMClientOptions,
  DBClient,
  DBClientOptions,
  SecretsClient,
  SecretsClientOptions,
} from '.'

export const ensureClient = <TClient>(context: ServiceContext, name: ClientNames): TClient => {
  const clients = (context.clients = context.clients || {})
  let client = clients[name]
  if (!client) {
    client = clients[name] = makeClient(context, name) as any
  }
  return client as TClient
}

const makeClient = (context: ServiceContext, name: ClientNames): Destroyable => {
  const { clientOptions } = context
  const options = clientOptions?.[name]

  if (!options) {
    throw InvalidService(name)
  }

  switch (name) {
    case 'ssm':
      return new SMClient(options as SMClientOptions, context)
    case 'db':
      return new DBClient(options as DBClientOptions, context)
    case 'secrets':
      return new SecretsClient(options as SecretsClientOptions, context)
  }
}
