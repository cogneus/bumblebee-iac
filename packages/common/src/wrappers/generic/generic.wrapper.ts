import 'source-map-support/register'
import { ServiceContext } from '../../types'
import { createWrapperContext } from '../utils'
import { GenericWrapper } from './types'
import { destroyClients } from '../../clients'

export const genericWrapper: GenericWrapper =
  ({ handler, getClientOptions, ...contextParams }) =>
  async (event, context: ServiceContext) => {
    const serviceContext = createWrapperContext({
      ...contextParams,
      context,
      logInfo: {
        event,
      },
      clientOptions: getClientOptions(),
    })
    const { log } = serviceContext

    try {
      log.debug('Process input', event as any)

      const result = await handler({
        event,
        serviceContext,
      })

      log.debug('Process complete', result as any)
      return result
    } catch (ex) {
      log.warn(`Process failed, ${ex.message}`)
      log.error(ex)
      throw ex
    } finally {
      destroyClients(serviceContext.clients)
    }
  }
