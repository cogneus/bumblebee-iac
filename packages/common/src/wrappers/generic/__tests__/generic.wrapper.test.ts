import { genericWrapper } from '../generic.wrapper'
import { ClientOptions } from '../../../clients'

const clientOptions: ClientOptions = {
  db: {
    maxAttempts: 3,
    region: 'eu-west-1',
    requestTimeout: 3000,
    tables: {
      test: {
        name: 'test-table-name',
        indices: {
          testIndex: { name: 'test-index-name' },
        },
        itemLimits: { defaultLimit: 20, maxLimit: 100 },
      },
    },
  },
}

describe('api gateway wrapper tests', () => {
  describe('should handle response, returns', () => {
    test('results matching snapshot', async () => {
      jest.resetAllMocks()
      const results: any[] = []
      const wrapper = genericWrapper({
        handler: async ({ serviceContext, event }) => {
          results.push({
            event,
            serviceContext,
          })
          return {
            body: '{}',
            statusCode: 200,
          }
        },
        region: 'eu-west-1',
        env: 'dev',
        logName: 'wrapper-test',
        serviceName: 'wrapper-test',
        getClientOptions: () => clientOptions,
      })

      const cb: any = null
      await wrapper({ test: 'event' }, {} as any, cb)
      expect(results).toMatchSnapshot()
    })
  })
})
