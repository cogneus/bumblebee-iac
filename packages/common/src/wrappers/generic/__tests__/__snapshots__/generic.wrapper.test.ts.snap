// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api gateway wrapper tests should handle response, returns results matching snapshot 1`] = `
[
  {
    "event": {
      "test": "event",
    },
    "serviceContext": {
      "clientOptions": {
        "db": {
          "maxAttempts": 3,
          "region": "eu-west-1",
          "requestTimeout": 3000,
          "tables": {
            "test": {
              "indices": {
                "testIndex": {
                  "name": "test-index-name",
                },
              },
              "itemLimits": {
                "defaultLimit": 20,
                "maxLimit": 100,
              },
              "name": "test-table-name",
            },
          },
        },
      },
      "env": "dev",
      "log": ConsoleLogger {
        "config": {
          "formatter": [Function],
          "includeSrc": true,
          "level": undefined,
          "levels": undefined,
        },
        "defaultEventCode": undefined,
        "formatter": [Function],
        "levels": {
          "debug": 2,
          "error": 5,
          "info": 3,
          "trace": 1,
          "warn": 4,
        },
        "minLevel": 3,
        "resolveLevel": [Function],
        "scope": {
          "env": "dev",
          "event": {
            "test": "event",
          },
          "name": "wrapper-test",
          "region": "eu-west-1",
          "serviceName": "wrapper-test",
        },
        "transformer": undefined,
      },
      "region": "eu-west-1",
    },
  },
]
`;
